django/cantusdb_project/cantusdb/settings.py

```diff
@@ -35,21 +36,38 @@
 # SECURITY WARNING: keep the secret key used in production secret!
 SECRET_KEY = os.getenv("CANTUSDB_SECRET_KEY")
 
-PROJECT_ENVIRONMENT = os.getenv("PROJECT_ENVIRONMENT")
+PROJECT_ENVIRONMENT: Optional[str] = os.getenv("PROJECT_ENVIRONMENT")
+
+if PROJECT_ENVIRONMENT not in ("DEVELOPMENT", "STAGING", "PRODUCTION"):
+    raise ValueError(
+        "The PROJECT_ENVIRONMENT environment variable must be either "
+        "DEVELOPMENT, STAGING, or PRODUCTION. "
+        f"Its current value is {PROJECT_ENVIRONMENT}."
+    )
+
+cantusdb_hosts: Optional[str] = os.getenv(f"CANTUSDB_HOSTS_{PROJECT_ENVIRONMENT}")
+cantusdb_origins: Optional[str] = os.getenv(f"CANTUSDB_ORIGINS_{PROJECT_ENVIRONMENT}")
+
+try:
+    ALLOWED_HOSTS: list = cantusdb_hosts.split(" ")
```

Comment:
You’re effectively checking the same condition 3 times.

Reply:
If you really want to do this (I think it's overkill -- there's a such thing as being too cautious, particularly because you're automating the deployment, but anyway...) then you should just check if it's null. This will anyway only catch the instances where you're trying to `split` on `None`. 

So:

```py
if cantusdb_hosts is None:
    raise ValueError(
        f"The PROJECT_ENVIRONMENT environment variable is set to {PROJECT_ENVIRONMENT}, "
        f"yet the CANTUSDB_HOSTS_{PROJECT_ENVIRONMENT} environment variable is not set."
    )

ALLOWED_HOSTS = cantusdb_hosts.split(" ")
```

Reply:
> You’re effectively checking the same condition 3 times.

I don't understand what you mean by this. Line 41, we check that `PROJECT_ENVIRONMENT` is properly set. Line 51, we check that the relevant `CANTUSDB_HOSTS_` variable is properly set, and analogously Line 54 and `CANTUSDB_ORIGINS_`.

> If you really want to do this (I think it's overkill -- there's a such thing as being too cautious, particularly because you're automating the deployment, but anyway...)

Fair enough. But if we expect misconfiguration errors to be caught by Ansible, then why should we bother handling a situation in our Python that will never arise? If we decide to prepare for this case that will never arise, isn't it worth setting something up something to aid future developers in debugging, rather than setting a safe fallback value which could obscure the cause of a problem?

Reply:
EAFP style has its place, but this probably isn't it. You know exactly what the types of `cantusdb_hosts` will be -- either `None` or a string. There's no need for a try/except if you know how to otherwise handle it. (Can you imagine trying to code where every `if` statement is a try/except? That would be a nightmare!)

(Part of me thinks EAFP style was useful back when Python didn't have any real concept of first-class types, so you could never really be sure what the type for any given variable was, and rather than check all possible types, just do something and catch the failure. But now we have type annotations, so you know exactly what it will be, and the type checker will tell you if you're not doing it right. So no need to do a try/except block.)

> But if we expect configuration errors to be caught by Ansible, then why should we bother handling a situation in our Python that will never arise?  If we decide to prepare for this case that will never arise, isn't it worth setting something up something to aid future developers in debugging, rather than setting a safe fallback value which could obscure the cause of a problem?

Indeed, that is the question. You're not coding an application for broad distribution, you have an install base of < 5, and you also control the end-to-end environment where these things get deployed. Automating the values in your settings files for staging / production and setting sane defaults for local development setups is the way I go.

Your "safe fallback" is the development environment. You can monkey around with those files all you like. But when it comes to a deployment, the rule is you *never* touch the config files by hand (Unless you're debugging a problem). If you need to make a change, make it in Ansible and deploy it. Want to provision a new server? Point Ansible at it and go. You should never get to a point where you're hand-deploying a server, and thus you'll never get to a point where these variables are not set. 

Your "future developers" will then just need to look at the Ansible deployment to see how the file is templated for staging / production.

Reply:
> (Can you imagine trying to code where every `if` statement is a try/except? That would be a nightmare!)

Sounds like [Hurl](https://ntietz.com/blog/introducing-hurl/) to me. Which is maybe a nightmare?